File: Sources/MonitorUI/Media/MonitorMediaFilterForm.swift
#if os(iOS)
    import SwiftUI

    /// Filter body shared by brand media pages. Native calendars stay in the shell.
    public struct MonitorMediaFilterForm: View {
        public struct ColorOption: Identifiable, Equatable, Sendable {
            public var id: UInt8
            public var label: String
            public init(id: UInt8, label: String) {
                self.id = id
                self.label = label
            }
        }

        public var formats: [String]
        public var resolutions: [String]
        public var colors: [ColorOption]
        public var formatSelection: Set<String>
        public var resolutionSelection: Set<String>
        public var colorSelection: Set<UInt8>
        public var dateStartLabel: String?
        public var dateEndLabel: String?
        public var hasDates: Bool
        public var onToggleFormat: (String) -> Void
        public var onToggleResolution: (String) -> Void
        public var onToggleColor: (UInt8) -> Void
        public var onPickStart: () -> Void
        public var onPickEnd: () -> Void
        public var onClear: () -> Void

        public init(
            formats: [String], resolutions: [String], colors: [ColorOption],
            formatSelection: Set<String>, resolutionSelection: Set<String>,
            colorSelection: Set<UInt8>, dateStartLabel: String?, dateEndLabel: String?,
            hasDates: Bool, onToggleFormat: @escaping (String) -> Void,
            onToggleResolution: @escaping (String) -> Void,
            onToggleColor: @escaping (UInt8) -> Void, onPickStart: @escaping () -> Void,
            onPickEnd: @escaping () -> Void, onClear: @escaping () -> Void
        ) {
            self.formats = formats
            self.resolutions = resolutions
            self.colors = colors
            self.formatSelection = formatSelection
            self.resolutionSelection = resolutionSelection
            self.colorSelection = colorSelection
            self.dateStartLabel = dateStartLabel
            self.dateEndLabel = dateEndLabel
            self.hasDates = hasDates
            self.onToggleFormat = onToggleFormat
            self.onToggleResolution = onToggleResolution
            self.onToggleColor = onToggleColor
            self.onPickStart = onPickStart
            self.onPickEnd = onPickEnd
            self.onClear = onClear
        }

        private var empty: Bool {
            formats.isEmpty && resolutions.isEmpty && !hasDates && colors.isEmpty
        }

        private var active: Bool {
            !formatSelection.isEmpty || !resolutionSelection.isEmpty || !colorSelection.isEmpty
                || dateStartLabel != nil || dateEndLabel != nil
        }

        public var body: some View {
            VStack(alignment: .leading, spacing: 10) {
                if !formats.isEmpty { chipSection("FORMAT", formats, formatSelection, onToggleFormat) }
                if !resolutions.isEmpty {
                    chipSection("RESOLUTION", resolutions, resolutionSelection, onToggleResolution)
                }
                if hasDates {
                    section("DATE") {
                        HStack(spacing: 6) {
                            dateField("Start", dateStartLabel, onPickStart)
                            Text("–").font(MonitorTheme.font(10.5, weight: .semibold))
                                .foregroundStyle(MonitorTheme.muted)
                            dateField("End", dateEndLabel, onPickEnd)
                        }
                    }
                }
                if !colors.isEmpty {
                    section("COLOUR") {
                        chipGrid(colors.map(\.label), Set(colors.filter { colorSelection.contains($0.id) }.map(\.label)))
                        { title in
                            if let option = colors.first(where: { $0.label == title }) {
                                onToggleColor(option.id)
                            }
                        }
                    }
                }
                if empty {
                    Text("Nothing in this tab to filter by.")
                        .font(MonitorTheme.font(11)).foregroundStyle(MonitorTheme.faint)
                }
                if active {
                    Button("Clear all filters", action: onClear)
                        .font(MonitorTheme.font(11, weight: .semibold))
                        .foregroundStyle(MonitorTheme.accent)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
        }

        private func section<Content: View>(_ title: String, @ViewBuilder content: () -> Content)
            -> some View
        {
            VStack(alignment: .leading, spacing: 5) {
                Text(title).font(MonitorTheme.font(9, weight: .bold)).foregroundStyle(MonitorTheme.muted)
                content()
            }
        }

        private func chipSection(
            _ title: String, _ titles: [String], _ active: Set<String>, _ toggle: @escaping (String) -> Void
        ) -> some View {
            section(title) { chipGrid(titles, active, toggle) }
        }

        private func chipGrid(
            _ titles: [String], _ active: Set<String>, _ toggle: @escaping (String) -> Void
        ) -> some View {
            let columns = [GridItem(.adaptive(minimum: 150), spacing: 5)]
            return LazyVGrid(columns: columns, spacing: 5) {
                ForEach(titles, id: \.self) { title in
                    Button {
                        toggle(title)
                    } label: {
                        Text(title)
                            .font(MonitorTheme.font(10, weight: .semibold))
                            .lineLimit(1).minimumScaleFactor(0.85)
                            .foregroundStyle(active.contains(title) ? MonitorTheme.accent : MonitorTheme.muted)
                            .frame(maxWidth: .infinity, minHeight: 30)
                            .padding(.horizontal, 8).padding(.vertical, 5)
                            .background(
                                active.contains(title)
                                    ? MonitorTheme.accent.opacity(0.14) : Color.white.opacity(0.04),
                                in: Capsule())
                    }
                    .buttonStyle(MonitorButtonStyle())
                }
            }
        }

        private func dateField(_ title: String, _ label: String?, _ action: @escaping () -> Void)
            -> some View
        {
            Button(action: action) {
                Text(label ?? title)
                    .font(MonitorTheme.font(10.5, weight: .semibold))
                    .foregroundStyle(label == nil ? MonitorTheme.muted : MonitorTheme.text)
                    .frame(maxWidth: .infinity, minHeight: 34)
                    .background(Color.white.opacity(0.04), in: RoundedRectangle(cornerRadius: 9))
            }
            .buttonStyle(MonitorButtonStyle())
            .accessibilityLabel(title)
            .accessibilityValue(label ?? "Any")
            .accessibilityIdentifier("monitor.media.filter.\(title.lowercased())")
        }
    }

#endif

File: Apps/Android/monitor-ui/src/main/kotlin/com/opencapture/monitorui/MonitorMediaFilterForm.kt
package com.opencapture.monitorui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.foundation.clickable
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp

@Immutable
data class MonitorMediaColorOption(val id: Int, val label: String)

@Composable
fun MonitorMediaFilterForm(
    formats: List<String>,
    resolutions: List<String>,
    colors: List<MonitorMediaColorOption>,
    formatSelection: Set<String>,
    resolutionSelection: Set<String>,
    colorSelection: Set<Int>,
    dateStartLabel: String?,
    dateEndLabel: String?,
    hasDates: Boolean,
    onToggleFormat: (String) -> Unit,
    onToggleResolution: (String) -> Unit,
    onToggleColor: (Int) -> Unit,
    onPickStart: () -> Unit,
    onPickEnd: () -> Unit,
    onClear: () -> Unit,
) {
    val empty = formats.isEmpty() && resolutions.isEmpty() && !hasDates && colors.isEmpty()
    val active = formatSelection.isNotEmpty() || resolutionSelection.isNotEmpty() ||
        colorSelection.isNotEmpty() || dateStartLabel != null || dateEndLabel != null
    Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
        if (formats.isNotEmpty()) FilterChips("FORMAT", formats, formatSelection, onToggleFormat)
        if (resolutions.isNotEmpty()) FilterChips("RESOLUTION", resolutions, resolutionSelection, onToggleResolution)
        if (hasDates) {
            FilterHeading("DATE")
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(6.dp),
            ) {
                FilterDateChip("Start", dateStartLabel, Modifier.weight(1f), onPickStart)
                Text("–", color = MonitorPalette.muted, style = MonitorTypography.text(10.5f, FontWeight.SemiBold))
                FilterDateChip("End", dateEndLabel, Modifier.weight(1f), onPickEnd)
            }
        }
        if (colors.isNotEmpty()) {
            FilterChips("COLOUR", colors.map { it.label }, colors.filter { it.id in colorSelection }.map { it.label }.toSet()) { title ->
                colors.firstOrNull { it.label == title }?.let { onToggleColor(it.id) }
            }
        }
        if (empty) {
            Text("Nothing in this tab to filter by.", color = MonitorPalette.faint, style = MonitorTypography.text(11f))
        }
        if (active) {
            Text(
                "Clear all filters",
                color = MonitorPalette.accent,
                style = MonitorTypography.text(11f, FontWeight.SemiBold),
                modifier = Modifier.padding(top = 2.dp).clickable(onClick = onClear),
            )
        }
    }
}

@Composable
private fun FilterHeading(title: String) {
    Text(title, color = MonitorPalette.muted, style = MonitorTypography.text(9f, FontWeight.Bold),
        modifier = Modifier.padding(bottom = 5.dp))
}

@Composable
private fun FilterChips(title: String, titles: List<String>, active: Set<String>, onToggle: (String) -> Unit) {
    Column(Modifier.padding(bottom = 2.dp), verticalArrangement = Arrangement.spacedBy(5.dp)) {
        FilterHeading(title)
        titles.forEach { title ->
            val on = title in active
            Text(
                title,
                color = if (on) MonitorPalette.accent else MonitorPalette.muted,
                style = MonitorTypography.text(10f, FontWeight.SemiBold),
                modifier = Modifier
                    .fillMaxWidth()
                    .clip(RoundedCornerShape(999.dp))
                    .background(if (on) MonitorPalette.accent.copy(alpha = .14f) else Color.White.copy(alpha = .04f))
                    .clickable { onToggle(title) }
                    .padding(horizontal = 8.dp, vertical = 8.dp),
            )
        }
    }
}

@Composable
private fun FilterDateChip(title: String, label: String?, modifier: Modifier, onClick: () -> Unit) {
    Text(
        label ?: title,
        color = if (label == null) MonitorPalette.muted else MonitorPalette.text,
        style = MonitorTypography.text(10.5f, FontWeight.SemiBold),
        modifier = modifier
            .clip(RoundedCornerShape(9.dp))
            .background(Color.White.copy(alpha = 0.04f))
            .clickable(onClick = onClick)
            .semantics { contentDescription = title }
            .heightIn(min = 34.dp)
            .padding(horizontal = 11.dp, vertical = 8.dp),
    )
}
